enhancements/sig-architecture/34-work-executor-group/README.md

```diff
@@ -0,0 +1,211 @@
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [website](https://github.com/open-cluster-management-io/website/)
+
+## Summary
+
+The proposal adds a new field `executorSubject` to the spec of `ManifestWork` 
+api, to support dynamic identity authentication/authorization upon the execution
+of `ManifestWork` resources. If the `executorSubject` is absent, it means the
+work agent will keep using the mounted service account to raise requests against
+the managed cluster which is also for preserving the backward compatibility.
+
+## Motivation
+
+### Separate the identity and its held permission for the work agent
+
+There can be different actors using the work api to distribute resource from
+the hub cluster. Currently the work agent has the cluster-admin access to the
+managed cluster which means that every hub users with write access to the
+work resources will be able to dispatch any resources to the managed cluster.
+This can be a security hole for OCM, e.g. one can apply/overwrite some
+cluster-scoped configuration to the managed cluster which potentially break/halt
+the cluster. Hence, we need a way to clarify the owner identity/role of the
+"ManifestWork" before it takes effect so that we can explicitly check whether 
+that owner has sufficient permission in the managed cluster.
+
+### Auditing the API request history from work agent
+
+Currently all the API requests raised by work agent in the managed cluster are
+from a fixed service account ("open-cluster-management-agent/klusterlet-work-sa"
+under the default installation). However in some cases, we need to clarify the
+source/owner identity of the `ManifestWork` resource so that we can record the
+api audits in a finer granularity. This will be helpful for the managed
+cluster's admin to be capable of knowing "who is operating that `ManifestWork`
+from the hub cluster".
+
+### Goals
+
+- `ManifestWork` resource be able to clarify the executor/owner identity before
+  actually writing the resources to the managed cluster.
+
+### Non-Goals
+
+- This proposal doesn't put any requirement to the selection of executor/owner in
+  the "ManifestWork", which allows hub admin to reference any valid Kubernetes 
+  user/role to be the executor/owner.
+
+## Proposal
+
+### API Changes
+
+This proposal adds a new field `executorSubject` to the `ManifestWork` api, 
+which is defaulted to nil if absent. A nil executor subject indicates that the 
+work agent doesn't perform any impersonation before requesting the hosting 
+managed cluster.
+
+
+A sample of the new `executorSubject` field in Yaml is shown below:
+
+```yaml
+apiVersion: work.open-cluster-management.io/v1
+kind: ManifestWork
+metadata: ...
+spec:
+  executorSubject:
+    type: [ServiceAccount|Username] # union discriminator for extensibility
+    serviceAccount:
+      name: ...
+      namespace: ...
```

Comment:
What stops me from specifying the username of a cluster admin and still applying any changes to the managed cluster?

Reply:
it is supposed to be defensed by an admission webhook, which is also elaborated in the proposal. the detail of admission webhook will not be covered in the current design, and it's always recommended to use `ServiceAccount` rather than `Username`. for `ServiceAccount` type executor, a feasible API model of the policy configuration of the admission webhook is `ManagedServiceAccountBinding`, e.g.:

```yaml
apiVersion: authentication.open-cluster-management.io/v1alpha1
kind: ManagedServiceAccountBinding
spec:
  ref:
    namespace: <namespace of the managed service-account>
    name: <name of the managed service-account>
  subjects:
    - kind: User
      name: dave
    - kind: Group
      name: my-group
    - kind: ServiceAccount
      namespace: <namespace>
      name: <name>
```

the binding api will define the mapping/accessibility between real hub users and the virtual service-account.

Reply:
If there is 1 or more `ManagedServiceAccountBinding` present, would you block the use of `executorSubject` for that ManagedCluster, or selectively applying the admission webhook (to ManagedCluster namespaces) for backwards compatibility?

Reply:
@jnpacker i think the webhook can do the selective applying via label selecting. e.g. the admission webhook only applies to those `ManifestWork` that has `work.open-cluster-management.io/executor-subject-gated: "true"`. so the existing works need to opt-in to the webhook by adding the label otherwise they're not protected, which keeps the backward compatibility. 

in abstract, by introducing the webhook, we will divide the `ManifestWork`'s write permission into two finer permissions:
- (1) the work executor who cannot editing the current executor subject but can modify all the other fields on the work. otherwise a hub user can unexpectedly leak the ownership of the `ManifestWork` and can't do anything to get it recovered.
- (2) the work owner who has complete access to the work e.g. create/delete the work, edit exeuctor subject 

the role (2) can be either a hub cluster global admin or a higher controller orchestrating the work across clusters. the role (1) can be an application domain owner. this is what will happen in detail when an arbitrary hub user trying to edit the `ManifestWork`:

1. does that hub user has the write permission upon the work in terms of Kubernetes RBAC? if no, rejected.
2. is that hub user writing upon a work resource w/ non-empty executor subject? if no, pass (for the backward compatibility).
3. is that hub user tweaking the current executor subject?
    - if yes, the hub user should have the RBAC permission to `manifestworks/assign` virtual subresource.
    - if no, fallthrough
4. is that hub user bound to the current executor subject? (which is checked by the admission webhook)
    - if the hub user has matching `ManagedServiceAccountBinding` bound, the edit passes.
    - if no, reject
    
 note that in the above process, the webhook is responsible for checking step 2 3 4.